packages/builders/src/messages/Message.ts

```diff
@@ -0,0 +1,544 @@
+import type { JSONEncodable } from '@discordjs/util';
+import type {
+	APIActionRowComponent,
+	APIAllowedMentions,
+	APIAttachment,
+	APIEmbed,
+	APIMessageActionRowComponent,
+	APIMessageReference,
+	APIPoll,
+	RESTPostAPIChannelMessageJSONBody,
+	Snowflake,
+} from 'discord-api-types/v10';
+import { ActionRowBuilder } from '../components/ActionRow.js';
+import { normalizeArray, type RestOrArray } from '../util/normalizeArray.js';
+import { resolveBuilder } from '../util/resolveBuilder.js';
+import { validate } from '../util/validation.js';
+import { AllowedMentionsBuilder } from './AllowedMentions.js';
+import { messagePredicate } from './Assertions.js';
+import { AttachmentBuilder } from './Attachment.js';
+import { MessageReferenceBuilder } from './MessageReference.js';
+import { EmbedBuilder } from './embed/Embed.js';
+import { PollBuilder } from './poll/Poll.js';
+
+export interface MessageBuilderData
+	extends Partial<
+		Omit<
+			RESTPostAPIChannelMessageJSONBody,
+			'allowed_mentions' | 'attachments' | 'components' | 'embeds' | 'message_reference' | 'poll'
+		>
+	> {
+	allowed_mentions?: AllowedMentionsBuilder;
+	attachments: AttachmentBuilder[];
+	components?: ActionRowBuilder[];
+	embeds: EmbedBuilder[];
+	message_reference?: MessageReferenceBuilder;
+	poll?: PollBuilder;
+}
+
+export class MessageBuilder implements JSONEncodable<RESTPostAPIChannelMessageJSONBody> {
+	/**
+	 * The API data associated with this message.
+	 */
+	private readonly data: MessageBuilderData;
+
+	/**
+	 * Gets the attachments of this message.
+	 */
+	public get attachments(): readonly AttachmentBuilder[] {
+		return this.data.attachments;
+	}
+
+	/**
+	 * Gets the components of this message.
+	 */
+	public get components(): readonly ActionRowBuilder[] {
+		return this.data.components ?? [];
```

Comment:
Why is this `?? []` but attachments/embeds aren't?

Reply:
This is typed as missing above in `MessageBuilderData`, so it seems the better question would be why it is missing in `MessageBuilderData` when the other two are not. `RESTPostAPIChannelMessageJSONBody` has all 3 potentially missing.